src/zenml/pipelines/run_utils.py

```diff
@@ -119,41 +120,53 @@ def deploy_pipeline(
     deployment: "PipelineDeploymentResponse",
     stack: "Stack",
     placeholder_run: Optional["PipelineRunResponse"] = None,
+    cleanup_placeholder_run: bool = False,
 ) -> None:
     """Run a deployment.
 
     Args:
         deployment: The deployment to run.
         stack: The stack on which to run the deployment.
         placeholder_run: An optional placeholder run for the deployment. This
-            will be deleted in case the pipeline deployment failed.
+            will be deleted in case the pipeline deployment failed and
+            `cleanup_placeholder_run` is set to True.
+        cleanup_placeholder_run: If True, the placeholder run will be deleted
+            in case the pipeline deployment failed.
 
     Raises:
         Exception: Any exception that happened while deploying or running
             (in case it happens synchronously) the pipeline.
     """
-    stack.prepare_pipeline_deployment(deployment=deployment)
+
+    def _cleanup_after_failure() -> None:
+        if not placeholder_run:
+            return
+
+        refreshed_run = Client().get_pipeline_run(
+            placeholder_run.id, hydrate=False
+        )
+        if refreshed_run.status != ExecutionStatus.INITIALIZING:
+            # The run is failed or some steps have already started
+            return
+
+        if cleanup_placeholder_run:
+            Client().delete_pipeline_run(placeholder_run.id)
+        else:
+            publish_failed_pipeline_run(placeholder_run.id)
 
     # Prevent execution of nested pipelines which might lead to
     # unexpected behavior
     previous_value = constants.SHOULD_PREVENT_PIPELINE_EXECUTION
     constants.SHOULD_PREVENT_PIPELINE_EXECUTION = True
     try:
+        stack.prepare_pipeline_deployment(deployment=deployment)
         stack.deploy_pipeline(
             deployment=deployment,
             placeholder_run=placeholder_run,
         )
     except Exception as e:
-        if (
-            placeholder_run
-            and Client().get_pipeline_run(placeholder_run.id).status
-            == ExecutionStatus.INITIALIZING
-        ):
-            # The run hasn't actually started yet, which means that we
-            # failed during initialization -> We don't want the
-            # placeholder run to stay in the database
-            Client().delete_pipeline_run(placeholder_run.id)
-
+        # TODO: Ideally this would not run if the orchestrator already created a run on their end?
```

Comment:
@bcdurak I remember after your PR with the orchestrator URLs, we ran into a case where the placeholder run was deleted but we would have wanted it to still be there, so we could see the URLs?
If I remember correctly, it was a syncronous orchestrator that failed to e.g. download the Docker image for the first step, which would lead to a failure where the logs URL would have been interesting.

Any idea how to best implement this? The best option I can come up with would be to separate the orchestrator into two different methods:
- one that does everything until the job is submitted to the orchestrator
- another one that waits in case the orchestrator is running sync

After the first method completes, we don't delete the placeholder run anymore.

Reply:
Or another suggestion: If the run has any metadata associated with it (e.g. the orchestrator URLs), we don't delete it anymore but instead just mark it as failed. That might be a much easier and quicker solution than my other suggestion, and will in most cases achieve the same thing?

Reply:
@bcdurak As discussed, we now don't delete the placeholder run anymore.